source/slang/type-system-shared.h

```diff
@@ -8,7 +8,12 @@ namespace Slang
 #define FOREACH_BASE_TYPE(X)    \
     X(Void)                     \
     X(Bool)                     \
```

Comment:
It seems as if Int and UInt here mean UInt32 and Int32 - but that is not clear by the naming convention. This is also confusing compared with the UInt and Int as used within the slang codebase which on 64 bit builds are 64 builds. 

If Int and UInt are really always 32 bits here then enum should reflect that. 

There is also perhaps an argument - depending on semantics for there to be a UInt/Int that represent 'word size'. The semantic difference can be around coercion and that coercing to 32 or 64 sizes being ok.

Reply:
Yes, the `BaseType::Int` and `::UInt` are 32 bit. The only reason I'm not renaming them as part of this change is that those names get used in various places and I didn't want to include the bulk rename of existing functionality in a change that was supposed to be adding new functionality. If I did that rename I'd probably also rename `Half`, `Float` and `Double` to `Float16`, `Float32`, and `Float64` just for consistency.

In the long term I'd probably like Slang to have distinct types `Int32`, `Int`, and `Int64`, where `Int` is conceptually "between" the 32- and 64-bit cases (so it is always safe to convert `Int32` to `Int` and `Int` to `Int64`). The IR would then be abstract about the "natural" integer size, and then we'd specialize it to something target-specific as part of lowering. Based on your comments I *think* we are on the same page about that.

(Actually, if you want to get extreme it might make sense to have all of `Int32`, `Int`, IntPtr`, and `Int64` where `IntPtr` (which might be named `Size` or something similar) is always the same size as a pointer on the target.)

Getting all of conversion and overloading to work correctly in the presence of such fine-grained types would be a challenge. Having *more* distinct integer types than C++ also makes it more challenging to generate C++ headers for Slang code, because two overloads that might be distinct in Slang might end up colliding in C++.